src/neo4j_genai/core/stores.py

```diff
@@ -0,0 +1,96 @@
+#  Copyright (c) "Neo4j"
+#  Neo4j Sweden AB [https://neo4j.com]
+#  #
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  #
+#      https://www.apache.org/licenses/LICENSE-2.0
+#  #
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""Result store interface
+and in-memory store implementation.
+"""
+
+from __future__ import annotations
+
+import abc
+from typing import Any
+
+from jsonpath_ng import parse
+
+
+class Store(abc.ABC):
+    """An interface to save component outputs"""
+
+    @abc.abstractmethod
+    def add(self, key: str, value: Any, overwrite: bool = True) -> None:
+        """
+        Args:
+            key (str): The key to access the data.
+            value (Any): The value to store in the data.
+            overwrite (bool): Whether to overwrite existing data.
+                If overwrite is False and the key already exists
+                in the store, an exception is raised.
+
+        Raises:
+            KeyError: If the key already exists in the store and overwrite is False.
+        """
+        pass
+
+    @abc.abstractmethod
+    def get(self, key: str) -> Any:
+        """Retrieve value for `key`.
+        If key not found, returns None.
+        """
+        pass
+
+    @abc.abstractmethod
+    def find_all(self, pattern: str) -> list[Any]:
+        """Find all values whose key matches pattern"""
+        pass
+
+    def all(self) -> dict[str, Any]:
```

Comment:
Why do we need all and empty methods here?

Reply:
`all` is used to return the results at the end, `empty` is used to clear the existing results if we ever re-run the same pipeline again (which I think can happen, for instance with different inputs)